### PR TITLE
Access Rules update

### PR DESCRIPTION
**Pull request by heylemur**
Summary of changes:
- Removed the granting access examples from the Managing_Accounts folders
- Deleted the Composite_Roles examples
- Added tags to the Managing_Resources examples
- Replaced the Managing_Roles examples with the new
- Added info on migrating to access rules to the Readme files

### CODE REVIEW

**Review by turner-strongdm — APPROVED** [review]
Good to go from my perspective.

**Review by jeffreylees — COMMENTED** [review]
The only things I see here are the same generic points from TF and Python PRs. Otherwise good to go.

File: README.md [hunk 1]
```diff
@@ -4,6 +4,14 @@ This is the examples repository for the [strongDM Go SDK](https://github.com/str
...
-> This repository is currently under construction. The content is expected to change in the coming weeks.
+> To increase flexibility when managing thousands of Resources, Role Grants have
+been deprecated in favor of Access Rules, which allow you to grant access based
+on Resource Tags and Type.
+>
+> Previously, you would grant a Role access to specific resources by ID via Role
+Grants. Now, when using Access Rules, the best practice is to grant Resources access based on
+Type and Tags.
+>
+>The following examples demonstrate Dynamic Access Rules with Tags and Resource Types, as well as Static Access Rules for backwards compatibility with Role Grants. If it is _necessary_ to grant access to specific Resources in the same way as RoleGrants did, you can use Resource IDs directly in Access Rules.
```
Comment (heylemur) [inline README.md:15]:
```suggestion
>The following examples demonstrate Dynamic Access Rules with tags and resource types, as well as Static Access Rules. If it is _necessary_ to grant access to specific Resources in the same way as Role Grants did, you can use Resource IDs directly in Access Rules.
```

File: README.md [hunk 2]
```diff
@@ -4,6 +4,14 @@ This is the examples repository for the [strongDM Go SDK](https://github.com/str
 
 ---
 > **NOTE:**  
-> This repository is currently under construction. The content is expected to change in the coming weeks.
+> To increase flexibility when managing thousands of Resources, Role Grants have
+been deprecated in favor of Access Rules, which allow you to grant access based
+on Resource Tags and Type.
+>
+> Previously, you would grant a Role access to specific resources by ID via Role
+Grants. Now, when using Access Rules, the best practice is to grant Resources access based on
```
Comment (heylemur) [inline README.md:12]:
```suggestion
Grants. Now, when using Access Rules, the best practice is to give Roles access to Resources based on 
```

File: README.md [hunk 3]
```diff
@@ -4,6 +4,14 @@ This is the examples repository for the [strongDM Go SDK](https://github.com/str
...
+Type and Tags.
```
Comment (heylemur) [inline README.md:13]:
```suggestion
type and tags.
```

File: README.md [hunk 4]
```diff
@@ -4,6 +4,14 @@ This is the examples repository for the [strongDM Go SDK](https://github.com/str
 
 ---
 > **NOTE:**  
-> This repository is currently under construction. The content is expected to change in the coming weeks.
+> To increase flexibility when managing thousands of Resources, Role Grants have
```
Comment (heylemur) [inline README.md:7]:
```suggestion
> To increase flexibility when managing a large volume of Resources, Role Grants have
```

File: 3_managing_roles/update_role/main.go [hunk 5]
```diff
@@ -41,42 +43,188 @@ func main() {
...
 	if err != nil {
-		log.Fatalf("Could not create role: %v", err)
+		log.Fatalf("failed to update role: %v", err)
 	}
 
-	roleID := createResponse.Role.ID
-	fmt.Println("Successfully created role.")
-	fmt.Println("\tID:", roleID)
+	// The RoleGrants API has been deprecated in favor of Access Rules.
+	// When using Access Rules, the best practice is to grant resources access based on type and tags.
```
Comment (heylemur) [inline 3_managing_roles/update_role/main.go:87]:
```suggestion
	// When using Access Rules, the best practice is to give Roles access to Resources based on type and tags.
```

File: 3_managing_roles/grant_access/main.go [hunk 6]
```diff
@@ -41,90 +43,165 @@ func main() {
...
 
-	roleID := roleResponse.Role.ID
-	fmt.Println("Successfully created role.")
-	fmt.Println("\tID:", roleID)
+	attachmentID := attachmentResponse.AccountAttachment.ID
+	fmt.Println("Successfully created account attachment.")
+	fmt.Println("\tID:", attachmentID)
+	
+	// The RoleGrants API has been deprecated in favor of Access Rules.
+	// When using Access Rules, the best practice is to grant resources access based on type and tags.
```
Comment (heylemur) [inline 3_managing_roles/grant_access/main.go:102]:
```suggestion
	// When using Access Rules, the best practice is to give Roles access to Resources based on type and tags.
```

File: 3_managing_roles/grant_access/main.go [hunk 7]
```diff
@@ -41,90 +43,165 @@ func main() {
...
 	}
 
-	roleID := roleResponse.Role.ID
-	fmt.Println("Successfully created role.")
-	fmt.Println("\tID:", roleID)
+	attachmentID := attachmentResponse.AccountAttachment.ID
+	fmt.Println("Successfully created account attachment.")
+	fmt.Println("\tID:", attachmentID)
+	
+	// The RoleGrants API has been deprecated in favor of Access Rules.
```
Comment (heylemur) [inline 3_managing_roles/grant_access/main.go:101]:
```suggestion
	// The Role Grants API has been deprecated in favor of Access Rules.
```

File: 3_managing_roles/revoke_access/main.go [hunk 8]
```diff
@@ -41,104 +43,188 @@ func main() {
...
+	_, err = client.Roles().Update(ctx, role)
 	if err != nil {
-		log.Fatalf("Could not create Postgres datasource: %v", err)
+		log.Fatalf("failed to update role: %v", err)
 	}
 
-	resourceID := resourceResponse.Resource.GetID()
-	fmt.Println("Successfully created Postgres datasource.")
-	fmt.Println("\tID:", resourceID)
+	// The RoleGrants API has been deprecated in favor of Access Rules.
```
Comment (heylemur) [inline 3_managing_roles/revoke_access/main.go:86]:
```suggestion
	// The Role Grants API has been deprecated in favor of Access Rules.
```

File: 3_managing_roles/update_role/main.go [hunk 9]
```diff
@@ -41,42 +43,188 @@ func main() {
...
+	_, err = client.Roles().Update(ctx, role)
 	if err != nil {
-		log.Fatalf("Could not create role: %v", err)
+		log.Fatalf("failed to update role: %v", err)
 	}
 
-	roleID := createResponse.Role.ID
-	fmt.Println("Successfully created role.")
-	fmt.Println("\tID:", roleID)
+	// The RoleGrants API has been deprecated in favor of Access Rules.
```
Comment (heylemur) [inline 3_managing_roles/update_role/main.go:86]:
```suggestion
	// The Role Grants API has been deprecated in favor of Access Rules.
```

File: 1_managing_resources/create_datasource/main.go [hunk 10]
```diff
@@ -53,9 +53,10 @@ func main() {
 		Password:     "example",
 		Database:     "example",
 		PortOverride: 19999,
+		Tags:		  "example=example",
```
Comment (heylemur) [inline 1_managing_resources/create_datasource/main.go:56]:
```suggestion
		Tags:		  "example=example"
```

File: 1_managing_resources/create_eks_cluster/main.go [hunk 11]
```diff
@@ -73,6 +73,7 @@ JrXnKV7Z4r9uWg==
 		ClusterName:          "example",
 		RoleArn:              "arn:aws:iam::000000000000:role/RoleName",
 		HealthcheckNamespace: "default",
+		Tags:				  "example=example",
```
Comment (heylemur) [inline 1_managing_resources/create_eks_cluster/main.go:76]:
```suggestion
		Tags:				  "example=example"
```

File: 1_managing_resources/delete_resource/main.go [hunk 12]
```diff
@@ -53,12 +53,13 @@ func main() {
 		Password:     "example",
 		Database:     "example",
 		PortOverride: 19999,
+		Tags:		  "example=example",
```
Comment (heylemur) [inline 1_managing_resources/delete_resource/main.go:56]:
```suggestion
		Tags:		  "example=example"
```

File: 1_managing_resources/update_resource/main.go [hunk 13]
```diff
@@ -53,9 +53,10 @@ func main() {
 		Password:     "example",
 		Database:     "example",
 		PortOverride: 19999,
+		Tags:		  "example=example",
```
Comment (heylemur) [inline 1_managing_resources/update_resource/main.go:56]:
```suggestion
		Tags:		  "example=example"
```

File: 3_managing_roles/create_role/main.go [hunk 14]
```diff
@@ -35,29 +37,33 @@ func main() {
...
+		// Create the client
+		client, err := sdm.New(
+			accessKey,
+			secretKey,
+		)
+		if err != nil {
+			log.Fatal("failed to create strongDM client:", err)
+		}
 
+	// Create a resource (e.g., Redis)
```
Comment (heylemur) [inline 3_managing_roles/create_role/main.go:49]:
```suggestion
	// Create a Resource (e.g., Redis)
```

File: 3_managing_roles/delete_role/main.go [hunk 15]
```diff
@@ -41,30 +43,34 @@ func main() {
 		secretKey,
 	)
 	if err != nil {
-		log.Fatalf("could not create client: %v", err)
+		log.Fatal("failed to create strongDM client:", err)
 	}
 
+	// Create a resource (e.g., Redis)
```
Comment (heylemur) [inline 3_managing_roles/delete_role/main.go:49]:
```suggestion
	// Create a Resource (e.g., Redis)
```

File: 3_managing_roles/grant_access/main.go [hunk 16]
```diff
@@ -41,90 +43,165 @@ func main() {
...
+	// When using Access Rules, the best practice is to grant resources access based on type and tags.
+	// If it is _necessary_ to grant access to specific resources in the same way as RoleGrants did,
+	// you can use resource IDs directly in Access Rules as shown in the following examples.
```
Comment (heylemur) [inline 3_managing_roles/grant_access/main.go:104]:
```suggestion
	// you can use Resource IDs directly in Access Rules as shown in the following examples.
```

File: 3_managing_roles/grant_access/main.go [hunk 17]
```diff
@@ -41,90 +43,165 @@ func main() {
...
+	}
+	return roleResp.Role.ID
+}
 
-	// Create a user
-	user := &sdm.User{
-		Email:     "example@example.com",
-		FirstName: "example",
-		LastName:  "example",
+// Example: Create a sample resource and return the ID
```
Comment (heylemur) [inline 3_managing_roles/grant_access/main.go:133]:
```suggestion
// Example: Create a sample Resource and return the ID
```

File: 3_managing_roles/grant_access/main.go [hunk 18]
```diff
@@ -41,90 +43,165 @@ func main() {
...
+	if err != nil {
+		return fmt.Errorf("error updating role: %v", err)
 	}
+	return nil
+}
 
-	attachmentResponse, err := client.AccountAttachments().Create(ctx, attachment)
+// Example: List Role grants via Access Rules
+func listRoleGrantsViaAccessRulesExample(ctx context.Context, client *sdm.Client) error {
+	// Create example resources
```
Comment (heylemur) [inline 3_managing_roles/grant_access/main.go:182]:
```suggestion
	// Create example Resources
```

File: 3_managing_roles/revoke_access/main.go [hunk 19]
```diff
@@ -41,104 +43,188 @@ func main() {
...
+	if err != nil {
+		return fmt.Errorf("error updating role: %v", err)
 	}
+	return nil
+}
 
-	accountResponse, err := client.Accounts().Create(ctx, user)
+// Example: Delete Role grant via Access Rules
+func deleteRoleGrantViaAccessRulesExample(ctx context.Context, client *sdm.Client) error {
+	// Create example resources
```
Comment (heylemur) [inline 3_managing_roles/revoke_access/main.go:167]:
```suggestion
	// Create example Resources
```

File: 3_managing_roles/revoke_access/main.go [hunk 20]
```diff
@@ -41,104 +43,188 @@ func main() {
...
-	fmt.Println("\tID:", accountID)
+	if len(role.AccessRules) != 1 || len(role.AccessRules[0].IDs) == 0 {
+		return fmt.Errorf("unexpected access rules in role")
+	}
 
-	// Assign account to role
-	attachment := &sdm.AccountAttachment{
-		AccountID: accountID,
-		RoleID:    roleID,
+	// Remove the ID of the second resource
```
Comment (heylemur) [inline 3_managing_roles/revoke_access/main.go:187]:
```suggestion
	// Remove the ID of the second Resource
```

File: 3_managing_roles/revoke_access/main.go [hunk 21]
```diff
@@ -41,104 +43,188 @@ func main() {
...
 
-	attachmentID := attachmentResponse.AccountAttachment.ID
-	fmt.Println("Successfully created account attachment.")
-	fmt.Println("\tID:", attachmentID)
-
-	// Detatch user from role
-	_, err = client.AccountAttachments().Delete(ctx, attachmentID)
+// Example: List Role grants via Access Rules
+func listRoleGrantsViaAccessRulesExample(ctx context.Context, client *sdm.Client) error {
+	// Create example resources
```
Comment (heylemur) [inline 3_managing_roles/revoke_access/main.go:205]:
```suggestion
	// Create example Resources
```

File: 3_managing_roles/update_role/main.go [hunk 22]
```diff
@@ -41,42 +43,188 @@ func main() {
 		secretKey,
 	)
 	if err != nil {
-		log.Fatalf("could not create client: %v", err)
+		log.Fatal("failed to create strongDM client:", err)
 	}
 
+	// Create a resource (e.g., Redis)
```
Comment (heylemur) [inline 3_managing_roles/update_role/main.go:49]:
```suggestion
	// Create a Resource (e.g., Redis)
```

File: 3_managing_roles/update_role/main.go [hunk 23]
```diff
@@ -41,42 +43,188 @@ func main() {
...
+		log.Fatalf("failed to update role: %v", err)
 	}
 
-	roleID := createResponse.Role.ID
-	fmt.Println("Successfully created role.")
-	fmt.Println("\tID:", roleID)
+	// The RoleGrants API has been deprecated in favor of Access Rules.
+	// When using Access Rules, the best practice is to grant resources access based on type and tags.
+	// If it is _necessary_ to grant access to specific resources in the same way as RoleGrants did,
+	// you can use resource IDs directly in Access Rules as shown in the following examples.
```
Comment (heylemur) [inline 3_managing_roles/update_role/main.go:89]:
```suggestion
	// you can use Resource IDs directly in Access Rules as shown in the following examples.
```

File: 3_managing_roles/update_role/main.go [hunk 24]
```diff
@@ -41,42 +43,188 @@ func main() {
...
 	}
-	fmt.Println("Successfully updated role.")
-	fmt.Println("\tID:", updateResponse.Role.ID)
-	fmt.Println("\tName:", updateResponse.Role.Name)
+	return nil
+}
+
+// Example: Delete Role grant via Access Rules
+func deleteRoleGrantViaAccessRulesExample(ctx context.Context, client *sdm.Client) error {
+	// Create example resources
```
Comment (heylemur) [inline 3_managing_roles/update_role/main.go:167]:
```suggestion
	// Create example Resources
```

File: 3_managing_roles/update_role/main.go [hunk 25]
```diff
@@ -41,42 +43,188 @@ func main() {
...
+	if err != nil {
+		return fmt.Errorf("error getting role: %v", err)
+	}
+	role := getResp.Role
+
+	if len(role.AccessRules) != 1 || len(role.AccessRules[0].IDs) == 0 {
+		return fmt.Errorf("unexpected access rules in role")
+	}
+
+	// Remove the ID of the second resource
```
Comment (heylemur) [inline 3_managing_roles/update_role/main.go:187]:
```suggestion
	// Remove the ID of the second Resource
```

File: 1_managing_resources/create_datasource/main.go [hunk 26]
```diff
@@ -53,9 +53,10 @@ func main() {
 		Password:     "example",
 		Database:     "example",
 		PortOverride: 19999,
+		Tags:		  "example=example"
```
Comment (heylemur) [inline 1_managing_resources/create_datasource/main.go:56]:
```suggestion
		Tags:     "example=example",
```

File: 1_managing_resources/create_eks_cluster/main.go [hunk 27]
```diff
@@ -73,6 +73,7 @@ JrXnKV7Z4r9uWg==
 		ClusterName:          "example",
 		RoleArn:              "arn:aws:iam::000000000000:role/RoleName",
 		HealthcheckNamespace: "default",
+		Tags:				  "example=example"
```
Comment (heylemur) [inline 1_managing_resources/create_eks_cluster/main.go:76]:
```suggestion
		Tags:              "example=example",
```

File: 1_managing_resources/create_rdp_server/main.go [hunk 28]
```diff
@@ -52,6 +52,7 @@ func main() {
 		Password:     "example",
 		Port:         3389,
 		PortOverride: 19999,
+		Tags:		  "example=example"
```
Comment (heylemur) [inline 1_managing_resources/create_rdp_server/main.go:55]:
```suggestion
		Tags:         "example=example",
```

File: 1_managing_resources/create_ssh_server/main.go [hunk 29]
```diff
@@ -50,6 +50,7 @@ func main() {
 		Hostname: "203.0.113.23",
 		Username: "example",
 		Port:     22,
+		Tags:	  "example=example"
```
Comment (heylemur) [inline 1_managing_resources/create_ssh_server/main.go:53]:
```suggestion
		Tags:	  "example=example",
```

File: 1_managing_resources/delete_resource/main.go [hunk 30]
```diff
@@ -53,12 +53,13 @@ func main() {
 		Password:     "example",
 		Database:     "example",
 		PortOverride: 19999,
+		Tags:		  "example=example"
```
Comment (heylemur) [inline 1_managing_resources/delete_resource/main.go:56]:
```suggestion
		Tags:         "example=example",
```

File: 1_managing_resources/update_resource/main.go [hunk 31]
```diff
@@ -53,9 +53,10 @@ func main() {
 		Password:     "example",
 		Database:     "example",
 		PortOverride: 19999,
+		Tags:		  "example=example"
```
Comment (heylemur) [inline 1_managing_resources/update_resource/main.go:56]:
```suggestion
		Tags:          "example=example",
```

**Review by jeffreylees — APPROVED** [review]

**Comment by heylemur** [comment]
@etodd thank you!

**Pull request merged**